anchor/anchorRecord.go

```diff
@@ -96,24 +113,49 @@ func (ar *AnchorRecord) MarshalAndSignV2(priv interfaces.Signer) ([]byte, []byte
 	return data, sig.Bytes(), nil
 }
 
-func (ar *AnchorRecord) Unmarshal(data []byte) error {
+// Non-exported, refactored function - is it better to use *string return for nil, or ""??
+func splitAnchorAndSignature(data []byte) (string, string, error) {
 	if len(data) == 0 {
-		return fmt.Errorf("Invalid data passed")
+		return "", "", fmt.Errorf("Invalid data passed")
 	}
 	str := string(data)
 	end := strings.LastIndex(str, "}}")
 	if end < 0 {
-		return fmt.Errorf("Found no closing bracket in `%v`", str)
+		return "", "", fmt.Errorf("Found no closing bracket in `%v`", str)
 	}
-	str = str[:end+2]
-	err := json.Unmarshal([]byte(str), ar)
+	// Signature comes after anchor
+	anchorStr := str[:end+2]
+	sigStr := str[end+2:]
+
+	return anchorStr, sigStr, nil
+}
+
+// Unmarshal unmarshals json format input into this AnchorRecord
+func (ar *AnchorRecord) Unmarshal(data []byte) error {
+	str, _, err := splitAnchorAndSignature(data)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal([]byte(str), ar)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// IsSame returns true iff all the fields of ar==ar2
```

Comment:
`(ar.Ethereum != nil && ar.Ethereum.IsSame(ar2.Ethereum) == false)`
So if `ar.Etheruem` is nil, and `ar2.Etheruem` is not nil, then this would report it as being the same?
If 1 has a nil, I'd suggest checking the other is also nil. This goes for the Bitcoin field too

Reply:
Good catch, I had intended that IsSame() would return proper false if nil is passed in, however, in looking at my implementation, it will crash. I think the correct fix is to change the bitcoin/ethereum IsSame() function to properly return false when nil is passed in. Right now it will crash. :-\

Reply:
I updated this as well to fix the nil input, and found a bad return

Reply:
I see how you resolved it.